### DISCUSSION TITLE
How am I supposed to convert a string to a Python list using `TypeAdapter`?

### DISCUSSION THREAD

**Discussion by madebylydia** [pydantic/pydantic#8393] [Question]
hello!

I am working with the `TypeAdapter` class to validate my variables and transform my data. I'd like to support as much types as possible so that's why I used Pydantic.
My issue is that I gather input from user, so I always get a string, no Python classes.

I wanted to to be able to accept a list of strings from the user, but it seems that neither of those work:
```py
TypeAdapter(list).validate_strings("1,2,3", strict=False)
# ValidationError: Input should be a valid array [type=list_type, input_value='1,2,3', input_type=str]
TypeAdapter(list).validate_strings("[1,2,3]", strict=False)
# ValidationError: Input should be a valid array [type=list_type, input_value='[1,2,3]', input_type=str]
TypeAdapter(list).validate_strings("['1', '2', '3']", strict=False)
# ValidationError: Input should be a valid array [type=list_type, input_value="['1', '2', '3']", input_type=str]
```
Unless I am doing something wrong here, I don't know if this is a bug or not. Is there any other way to achieve what I'm trying? Thank you!

(also planning to adapt dict, so I believe it would behave the same way, let's start with lists first)

**Comment by a-gertner** [comment]
Was just looking at the same problem.

```Python
TypeAdapter(List[str]).validate_json('["1", "2", "3"]', strict=False)
['1', '2', '3']
TypeAdapter(List[int]).validate_json('[1, 2, 3]', strict=False)
[1, 2, 3]
```
Taken from here: https://docs.pydantic.dev/latest/concepts/strict_mode/#type-coercions-in-strict-mode

**Comment by oliverlambson** [comment]
Your example data has 3 different formats:

1. `"[1,2,3]"` --> valid json
2. `"1,2,3"` --> comma separated
3. `"['1', '2', '3']"` --> not json because of single-quoted strings

## If your data is always one of the formats

### 1. If your input data format is always json, it's easy:

```python
from typing import Any
from pydantic import TypeAdapter, Json

TypeAdapter(Json[list[int]]).validate_strings('[1, 2]')
# [1, 2]
TypeAdapter(Json[list[int]]).validate_strings('["1", "2"]')
# [1, 2]
TypeAdapter(Json[list[Any]]).validate_strings('[1, "a"]')
# [1, 'a']
```

### 2. If your input data is always csv, you'll need a custom `BeforeValidator`:

```python
from typing import Annotated, Any
from pydantic import BeforeValidator, TypeAdapter

def csv_to_list(value: Any) -> list[str] | Any:
    if isinstance(value, str):
        return value.split(",")
    raise ValueError("Not a csv string")
    
type CsvList[T] = Annotated[list[T], BeforeValidator(csv_to_list)]
TypeAdapter(CsvList[int]).validate_strings("1,2,3")
# [1, 2, 3]
```

### 3. If your data format is any other custom format, you'll also need a `BeforeValidator`

Adapt the csv code in (2) with a custom validation function.

## If your data is any of the formats, use a `Union`

```python
from typing import Annotated, Any
from pydantic import BeforeValidator, TypeAdapter, Json

def csv_to_list(value: Any) -> list[str] | Any:
    if isinstance(value, str) and not value.startswith("["): # stricter validation
        return value.split(",")
    raise ValueError("Not a csv string")
    
type CsvList[T] = Annotated[list[T], BeforeValidator(csv_to_list)]

type FlexiList[T] = Json[list[Any]] | CsvList[Any] # Union of possible types

TypeAdapter(FlexiList[int]).validate_strings("1,2,3")
# [1, 2, 3]
TypeAdapter(FlexiList[int]).validate_strings('["1", "2", "3"]')
# [1, 2, 3]
TypeAdapter(FlexiList[Any]).validate_strings('[1, "a"]')
# [1, 'a']
```

---

> [!NOTE]
> Generally it's considered preferable to push your validation up towards the user and restrict the input they can provide, so I would only go down this route if you can't do that)
